language-plutus-core/stdlib/Language/PlutusCore/StdLib/Type.hs

```diff
@@ -1,43 +1,170 @@
+    k <- freshTyName ann "k"
+    args <- for (zip [1 :: Int ..] argKinds) $ \(argN, argKind) -> do
+        name <- freshTyNameText ann $ "a" <> prettyText argN
+        return $ TyVarDecl ann name argKind
+    spine <- getSpine ann $ map tyDeclVar args
+
+    return
+        . TyLam ann k (KindArrow ann (KindArrow ann (toRecKind ann argKinds) $ Type ann) $ Type ann)
+        . mkIterTyLam ann args
+        . TyApp ann (TyVar ann k)
+        $ spine
+
+packagePatternBodyN
+    :: (Kind () -> Quote (Type TyName ()))
+       -- ^ Some type-level function that receives @withSpine@ and a pattern functor that binds
+       -- @n + 1@ type variables (where @1@ represents the variable responsible for recursion)
+       -- using type-level lambdas.
+    -> ann              -- ^ An annotation placed everywhere we do not have annotations.
+    -> TyName ann       -- ^ The name for the @1@ varible responsible for recursion.
+    -> [Kind ann]       -- ^ A list of kinds for @n@ type variables.
+    -> Type TyName ann  -- ^ The body of a pattern functor that consequently binds type variables.
+    -> Quote (Type TyName ann)
+packagePatternBodyN getFun ann name argKinds patBody = do
+    withSpine <- getWithSpine ann argKinds
+    let pat = TyLam ann name (toRecKind ann argKinds) patBody
+    fun <- getFun . void $ toRecKind ann argKinds
+    return $ mkIterTyApp ann (ann <$ fun) [withSpine, pat]
+
+getTyFix :: ann -> TyName ann -> [Kind ann] -> Type TyName ann -> Quote (Type TyName ann)
+getTyFix = packagePatternBodyN getTyFixN
+
+getWrap
+    :: ann
+    -> TyName ann
+    -> [Kind ann]
+    -> Type TyName ann
+    -> Quote ([Type TyName ann] -> Term TyName Name ann -> Term TyName Name ann)
+getWrap ann name argKinds patBody = do
+    pat1 <- packagePatternBodyN getToPatternFunctor ann name argKinds patBody
+    toSpine <- getToSpine ann
+    -- TODO: check lengths match.
+    return $ IWrap ann pat1 . toSpine . zipWith (flip $ TyDecl ann) argKinds
 
 -- | A 'Type' that starts with a 'TyFix' (i.e. a recursive type) packaged along with a
 -- specified 'Wrap' that allows to construct elements of this type.
-data RecursiveType tyname a = RecursiveType
-    { _recursiveWrap :: forall name. Term tyname name a -> Term tyname name a
-    , _recursiveType :: Type tyname a
+data RecursiveType ann = RecursiveType
+    { _recursiveWrap :: [Type TyName ann] -> Term TyName Name ann -> Term TyName Name ann
+    , _recursiveType :: Type TyName ann
     }
 
--- | Apply a 'HoledType' to a 'Type'.
-holedTyApp :: HoledType tyname () -> Type tyname () -> HoledType tyname ()
-holedTyApp (HoledType name cont) arg = HoledType name $ \hole -> TyApp () (cont hole) arg
-
--- | Convert a 'HoledType' to the corresponding 'RecursiveType'.
-holedToRecursive :: HoledType tyname () -> RecursiveType tyname ()
-holedToRecursive (HoledType name cont) =
-    RecursiveType (Wrap () name $ cont id) (cont $ TyFix () name)
+makeRecursiveType
+    :: ann
+    -> TyName ann
+    -> [Kind ann]
```

Comment:
Could this be a list of `TyVarDecl`s? I'd like to be able to name the final type lambda arguments if possible, just to make it easier to comprehend.

Reply:
Sure.

Reply:
Also, perhaps it would be nice to encapsulate the common case where the pattern functor is precisely a type lambda expecting the same arguments as the overall thing. So perhaps a version of this that takes `TyVarDecl`s and also binds them at the start of the pattern functor for you?

Reply:
I tried doing this, but it was kinda ugly to my taste. I want to do another thing: pass lambdas as they are passed now, strip them, turn the variables into a list of `TyVarDecl` and use that. This is ugly from the implementation perspective, but I like how it looks to the user.
Though, perhaps for you it's nicer to just pass lists of `TyVarDecl`? If so, I can implement that.

Reply:
Well, what I want is for the lambda outside the `ifix` to have the same name variable names as the lambda inside the `ifix`, because that corresponds to a type variable name that a human actually wrote. 

I don't really mind how we do this. I'm happy to pass around lists of `TyVarDecl`s because I've already got them!

I don't like the idea of stripping the lamdas off the front - that assumes that you want all of those variables to be included. I can't immediately think of a time this might do the wrong thing, but it feels risky to me.